### PR TITLE
Expose delta targeting, early exits, and commissions in AI agent tools

### PR DESCRIPTION
**Pull request by michaelchu**
The core library added per-leg delta targeting, P&L-based early exits,
and commission modeling but the AI chat agent tools were not updated
to expose these parameters. The LLM could not use them.

Changes:
- Add leg1_delta through leg4_delta, delta_interval, stop_loss,
  take_profit, max_hold_days, and commission to StrategyParamsMixin
  (propagates to run_strategy, simulate, and scan_strategies tools)
- Add DeltaTarget model for nested {target, min, max} schema
- Update system prompt with delta targeting table, early exit docs,
  and commission parameter documentation
- Update tool descriptions for run_strategy, suggest_strategy_params,
  and check_data_quality to reference delta targeting
- Update suggest_strategy_params to recommend per-leg delta targets
  based on strategy type (iron condor, butterfly, spread, etc.)

https://claude.ai/code/session_019jiM2z4MBrKKuSv5qDdF5m

### CODE REVIEW

File: optopsy/ui/tools/_schemas.py [hunk 1]
```diff
@@ -649,7 +649,8 @@ def get_required_option_type(strategy_name: str) -> str | None:
     "check_data_quality": (
         "Run an integrity check on a loaded dataset against the columns "
         "and data quality the backtesting engine requires. Reports "
-        "missing/mistyped required columns, optional column availability "
+        "missing/mistyped required columns (including delta, which is "
+        "required for strike selection), optional column availability "
         "(Greeks, volume, IV), null analysis on critical columns, bid/ask "
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_schemas.py:654]: Good catch — the wording was misleading. Updated the description to clarify that delta is an optional column needed specifically for delta-based strike selection, not a hard-required column for the engine overall. The check_data_quality implementation correctly treats it as optional under Greeks, and the description now matches.

File: optopsy/ui/tools/_data_inspector.py [hunk 2]
```diff
@@ -228,14 +228,22 @@ def _handle_suggest_strategy_params(
...
         delta_stats = {}
 
+    # Default delta targets per strategy type (unsigned absolute values)
+    _DEFAULT_DELTA = {"target": 0.30, "min": 0.20, "max": 0.40}
+    _DEFAULT_ATM_DELTA = {"target": 0.50, "min": 0.40, "max": 0.60}
+    _DEFAULT_OTM_DELTA = {"target": 0.10, "min": 0.05, "max": 0.20}
+    _DEFAULT_ITM_WING_DELTA = {"target": 0.40, "min": 0.30, "max": 0.50}
+    _DEFAULT_OTM_WING_DELTA = {"target": 0.10, "min": 0.05, "max": 0.20}
+    _DEFAULT_DEEP_ITM_DELTA = {"target": 0.80, "min": 0.60, "max": 0.95}
+
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_data_inspector.py]: Agreed — fixed. When delta data is unavailable, the tool now:
1. Omits per-leg delta target recommendations from the output
2. Includes an explicit warning that delta-based strike selection won't work and the engine will fall back to OTM% filtering

[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_data_inspector.py]: Good point — replaced the local constant declarations with imports from `optopsy.strategies._helpers`, which is the single source of truth for these defaults. This eliminates the duplication and drift risk.

File: optopsy/ui/tools/_models.py [hunk 3]
```diff
@@ -104,6 +104,20 @@ class CombineMode(str, Enum):
...
+
+    target: float = Field(
+        ..., gt=0, le=1, description="Target absolute delta (e.g. 0.30)"
+    )
+    min: float = Field(
+        ..., gt=0, le=1, description="Minimum absolute delta (e.g. 0.20)"
+    )
+    max: float = Field(
+        ..., gt=0, le=1, description="Maximum absolute delta (e.g. 0.40)"
+    )
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_models.py:129]: Agreed — added a `@model_validator(mode="after")` that enforces `min <= target <= max`, plus `model_config = {"extra": "forbid"}` to reject unknown keys. Invalid ranges now fail fast with a clear validation error.

File: optopsy/ui/tools/_models.py [hunk 4]
```diff
@@ -117,6 +131,73 @@ class StrategyParamsMixin(BaseModel):
...
+            "regardless of DTE or P&L. Omit for no time-based exit."
+        ),
+    )
+    commission: float | None = Field(
+        None,
+        description=(
+            "Commission per contract in dollars (e.g. 0.65). "
+            "Applied to each leg at entry and exit. Omit for no commissions."
+        ),
+    )
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_models.py:225]: Fixed — added Pydantic field constraints matching the documented behavior:
- `delta_interval`: `gt=0, le=1`
- `stop_loss`: `lt=0`
- `take_profit`: `gt=0`
- `max_hold_days`: `gt=0`
- `commission`: `ge=0`

Invalid tool inputs will now be rejected at the validation boundary before reaching the strategy engine.

File: tests/test_tools_models.py [hunk 5]
```diff
@@ -187,6 +188,155 @@ def test_list_results(self):
...
+    def test_rejects_missing_fields(self):
+        with pytest.raises(ValidationError):
+            DeltaTarget.model_validate({"target": 0.30})
+
+    def test_dumps_to_dict(self):
+        d = DeltaTarget.model_validate({"target": 0.30, "min": 0.20, "max": 0.40})
+        dumped = d.model_dump()
+        assert isinstance(dumped, dict)
+        assert dumped == {"target": 0.30, "min": 0.20, "max": 0.40}
+
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply tests/test_tools_models.py:239]: Good call — added three test cases covering the validation boundary conditions:
- `test_rejects_min_greater_than_target`: asserts `min > target` is rejected with a clear error message
- `test_rejects_target_greater_than_max`: asserts `target > max` is rejected
- `test_rejects_extra_fields`: verifies that `extra: 'forbid'` config rejects unknown keys

All three use `pytest.raises(ValidationError, match=...)` to confirm the specific error messages from the `@model_validator`.

File: tests/test_tools_check_data_quality.py [hunk 6]
```diff
@@ -16,7 +16,7 @@
 
 @pytest.fixture
 def clean_data():
-    """Minimal clean option dataset — all 7 required columns, no issues."""
+    """Minimal clean option dataset — all 8 required columns, no issues."""
     exp = datetime.datetime(2024, 3, 15)
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply tests/test_tools_check_data_quality.py:20]: Good catch — you're right that the engine's `expected_types` in `checks.py` includes `underlying_price`. Updated the fixture docstring and all assertions to say "9 required columns" and added `underlying_price` to `_CORE_REQUIRED_COLS`. The test `test_underlying_price_not_required` has been renamed to `test_underlying_price_required` and now asserts that dropping the column triggers a FAIL.

File: optopsy/ui/tools/_quality_checks.py [hunk 7]
```diff
@@ -33,6 +33,7 @@ def _handle_check_data_quality(arguments, dataset, signals, datasets, results, _
         "strike": ("int64", "float64"),
         "bid": ("int64", "float64"),
         "ask": ("int64", "float64"),
+        "delta": ("int64", "float64"),
     }
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_quality_checks.py:38]: Fixed — `underlying_price` is now included in `_CORE_REQUIRED_COLS` alongside `delta`, bringing the total to 9 required columns. This matches the engine's `expected_types` in `checks.py` exactly. Also removed `underlying_price` from the optional columns list since it's no longer optional. All count references (PASS messages, test assertions, fixture docstrings) updated accordingly.

File: optopsy/ui/tools/_schemas.py [hunk 8]
```diff
@@ -650,7 +650,8 @@ def get_required_option_type(strategy_name: str) -> str | None:
         "Run an integrity check on a loaded dataset against the columns "
         "and data quality the backtesting engine requires. Reports "
         "missing/mistyped required columns, optional column availability "
-        "(Greeks, volume, IV), null analysis on critical columns, bid/ask "
+        "(Greeks including delta — needed for delta-based strike selection, "
+        "volume, IV), null analysis on critical columns, bid/ask "
         "quality (zero bids, crossed markets, spread stats), date coverage "
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_schemas.py:655]: Fixed — updated the description to explicitly list delta and underlying_price as required columns: `"missing/mistyped required columns (including delta and underlying_price)"`. The parenthetical about delta being optional under Greeks has been removed since it's now correctly treated as required in the implementation.

File: optopsy/ui/tools/_data_inspector.py [hunk 9]
```diff
@@ -244,20 +255,91 @@ def _handle_suggest_strategy_params(
...
+    if not has_delta and strategy_name:
+        delta_warning = (
+            "WARNING: Dataset has no delta column. Delta-based strike selection "
+            "will not work. To use delta targeting, load data that includes "
+            "a 'delta' column."
+        )
+        strategy_note = (
+            f"{delta_warning} {strategy_note}" if strategy_note else delta_warning
+        )
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_data_inspector.py:376]: Good point — the warning now distinguishes between the two cases:

1. **Column missing**: `"Dataset has no delta column..."`
2. **Column present but all-null**: `"Delta column exists but contains no usable (non-null) values..."`

The logic already had the distinction available (the code checks `"delta" in active_ds.columns` before calling `.dropna()`), it just wasn't surfaced in the message.

File: optopsy/ui/tools/_quality_checks.py [hunk 10]
```diff
@@ -55,19 +57,18 @@ def _handle_check_data_quality(arguments, dataset, signals, datasets, results, _
             f"**Required Columns** — dtype mismatches: {'; '.join(dtype_mismatches)}\n"
         )
     else:
-        findings.append("PASS: all 7 required columns present with correct dtypes")
+        findings.append("PASS: all 9 required columns present with correct dtypes")
         display_parts.append(
-            "**Required Columns** — all 7 present with correct dtypes\n"
+            "**Required Columns** — all 8 present with correct dtypes\n"
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_quality_checks.py]: Good catch — missed the display message when updating the llm_summary. Fixed to say "all 9 present" in both places now.

File: optopsy/datafeeds.py [hunk 11]
```diff
@@ -8,7 +8,7 @@
 The module also handles:
 - Date column inference (``_infer_date_cols``)
 - Optional date-range filtering (``_trim_dates``)
-- Optional Greek and liquidity columns (delta, gamma, volume, etc.)
+- Optional Greek and liquidity columns (gamma, theta, vega, volume, etc.)
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/datafeeds.py]: Updated — the module docstring now clarifies that delta is required and only the other Greeks/liquidity columns are optional.

File: optopsy/ui/tools/_data_inspector.py [hunk 12]
```diff
@@ -244,20 +255,98 @@ def _handle_suggest_strategy_params(
             "back_dte_max": min(120, dte_stats["p90"]),
         }
         strategy_note = (
-            "Calendar strategy — use front/back DTE instead of max_entry_dte."
+            "Calendar strategy — use front/back DTE instead of max_entry_dte. "
+            "Delta defaults apply per-leg automatically."
         )
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_data_inspector.py]: Good point — made the calendar strategy note conditional on `has_delta`. When delta data is unavailable, the note now omits the "Delta defaults apply per-leg automatically" sentence to avoid implying delta-based selection will work.

File: optopsy/ui/tools/_quality_checks.py [hunk 13]
```diff
@@ -76,8 +77,6 @@ def _handle_check_data_quality(arguments, dataset, signals, datasets, results, _
 
     if available_optional:
         features: list[str] = []
-        if "delta" in available_optional:
-            features.append("delta filtering supported")
         if "implied_volatility" in available_optional:
             features.append("IV surface/signals available")
         if "volume" in available_optional:
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_quality_checks.py:82]: Good catch — removed the unreachable `if "underlying_price" in available_optional` branch since `underlying_price` is no longer in `_OPTIONAL_COLS`.

File: tests/test_tools_check_data_quality.py [hunk 14]
```diff
@@ -140,6 +139,7 @@ def test_no_optional_columns(self):
                 "strike": [480.0],
                 "bid": [5.0],
                 "ask": [5.10],
+                "delta": [0.50],
             }
         )
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply tests/test_tools_check_data_quality.py:145]: Fixed — added `underlying_price` to the inline DataFrame so the fixture actually has all 9 required columns. Test now correctly exercises the "no optional columns" path without masking a missing-required-column failure.

File: tests/test_tools_suggest_params.py [hunk 15]
```diff
@@ -204,6 +215,114 @@ def test_delta_distribution_shown(self, option_data):
...
+        """Reverse iron condor shares iron condor delta recommendations."""
+        result = execute_tool(
+            "suggest_strategy_params",
+            {"strategy_name": "reverse_iron_condor"},
+            wide_dte_data,
+        )
+        reco = _extract_recommended_json(result.user_display)
+        assert reco is not None
+        assert reco["leg1_delta"]["target"] == 0.10
+        assert reco["leg2_delta"]["target"] == 0.30
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply tests/test_tools_suggest_params.py:324]: Good catch — added assertions for `leg3_delta.target == 0.30` and `leg4_delta.target == 0.10` to match the iron_condor test pattern and fully cover all four legs.

File: tests/test_tools_suggest_params.py [hunk 16]
```diff
@@ -204,6 +215,114 @@ def test_delta_distribution_shown(self, option_data):
...
+
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply tests/test_tools_suggest_params.py:339]: Added `test_protective_put_recommends_deep_itm_plus_otm` — mirrors the covered_call test and verifies `leg1_delta.target == 0.80` and `leg2_delta.target == 0.30`.

File: optopsy/ui/tools/_quality_checks.py [hunk 17]
```diff
@@ -27,12 +27,14 @@ def _handle_check_data_quality(arguments, dataset, signals, datasets, results, _
...
         "underlying_symbol": ("object", "str"),
+        "underlying_price": ("int64", "float64"),
         "option_type": ("object", "str"),
         "expiration": ("datetime64[ns]", "datetime64[us]"),
         "quote_date": ("datetime64[ns]", "datetime64[us]"),
         "strike": ("int64", "float64"),
         "bid": ("int64", "float64"),
         "ask": ("int64", "float64"),
+        "delta": ("int64", "float64"),
     }
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_quality_checks.py:38]: Good point — added `delta` and `underlying_price` to `_NULL_CHECK_COLS` so the null analysis section now covers required numeric columns. If either is entirely null, the tool will report a WARN with the null percentage, preventing a false PASS for delta targeting readiness.

File: optopsy/ui/tools/_models.py [hunk 18]
```diff
@@ -104,6 +104,30 @@ class CombineMode(str, Enum):
...
+    model_config = {"extra": "forbid"}
+
+    target: float = Field(
+        ..., gt=0, le=1, description="Target absolute delta (e.g. 0.30)"
+    )
+    min: float = Field(
+        ..., gt=0, le=1, description="Minimum absolute delta (e.g. 0.20)"
+    )
+    max: float = Field(
+        ..., gt=0, le=1, description="Maximum absolute delta (e.g. 0.40)"
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_models.py]: Good catch on the inconsistency — updated the docstring and field descriptions to say "(0, 1] range" to accurately reflect the `gt=0, le=1` validators. Zero delta doesn't make practical sense for strike selection, so keeping `gt=0` is correct.

File: tests/test_tools_suggest_params.py [hunk 19]
```diff
@@ -204,6 +215,128 @@ def test_delta_distribution_shown(self, option_data):
...
+        result = execute_tool(
+            "suggest_strategy_params",
+            {"strategy_name": "protective_put"},
+            option_data,
+        )
+        reco = _extract_recommended_json(result.user_display)
+        assert reco is not None
+        assert reco["leg1_delta"]["target"] == 0.80
+        assert reco["leg2_delta"]["target"] == 0.30
+
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply tests/test_tools_suggest_params.py:339]: Added 4 test cases covering the no-delta scenarios:
- `test_no_delta_column_omits_delta_recommendations`: missing delta column, verifies no leg*_delta keys and WARNING present
- `test_all_null_delta_omits_delta_recommendations`: all-null delta column, same checks plus "no usable" message
- `test_no_delta_strategy_note_omits_delta_values`: verifies the strategy note doesn't mention specific delta values when delta is unavailable
- `test_reverse_iron_butterfly_same_as_iron_butterfly`: added for completeness

[the remark this answers is not in the document]
Comment (michaelchu) [reply tests/test_tools_suggest_params.py:339]: Added `test_reverse_iron_butterfly_same_as_iron_butterfly` — verifies all four legs match the iron_butterfly defaults (outer at 0.10, inner at 0.50 ATM).

File: optopsy/ui/tools/_data_inspector.py [hunk 20]
```diff
@@ -228,36 +237,121 @@ def _handle_suggest_strategy_params(
...
+        recommended["max_entry_dte"] = min(45, dte_stats["p75"])
+        if has_delta:
+            recommended["leg1_delta"] = _DEFAULT_OTM_DELTA
+            recommended["leg2_delta"] = _DEFAULT_DELTA
+            recommended["leg3_delta"] = _DEFAULT_DELTA
+            recommended["leg4_delta"] = _DEFAULT_OTM_DELTA
+        strategy_note = (
+            "Iron condor — outer legs at 0.10 delta (wings), inner legs at 0.30 delta. "
+            "Typically works best in the 20-45 DTE range."
+        )
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_data_inspector.py]: Fixed — all strategy notes are now conditional on `has_delta`. When delta data is unavailable, notes focus on DTE range or say "uses OTM% filtering without delta data" instead of mentioning specific delta values. This prevents the confusing combination of a delta warning followed by delta-specific recommendations.

File: optopsy/ui/tools/_data_inspector.py [hunk 21]
```diff
@@ -228,36 +237,121 @@ def _handle_suggest_strategy_params(
...
+        if has_delta:
+            recommended["leg1_delta"] = _DEFAULT_OTM_DELTA
+            recommended["leg2_delta"] = _DEFAULT_ATM_DELTA
+            recommended["leg3_delta"] = _DEFAULT_ATM_DELTA
+            recommended["leg4_delta"] = _DEFAULT_OTM_DELTA
         strategy_note = (
-            "Multi-leg strategies typically work best in the 20-45 DTE range."
+            "Iron butterfly — outer legs at 0.10 delta, inner legs at 0.50 (ATM). "
+            "Typically works best in the 20-45 DTE range."
+        )
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_data_inspector.py]: Fixed alongside the iron_condor case — iron_butterfly note is now conditional on `has_delta` with a DTE-only fallback message.

File: optopsy/ui/tools/_data_inspector.py [hunk 22]
```diff
@@ -228,36 +237,121 @@ def _handle_suggest_strategy_params(
...
+            recommended["leg1_delta"] = _DEFAULT_DEEP_ITM_DELTA
+            recommended["leg2_delta"] = _DEFAULT_DELTA
+        strategy_note = (
+            "Protective put — deep ITM put (0.80 delta) + 0.30 delta long put."
+        )
+    elif strategy_name:
+        # Single-leg strategies
+        if has_delta:
+            recommended["leg1_delta"] = _DEFAULT_DELTA
+        strategy_note = "Default delta target: 0.30 (range 0.20-0.40)."
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_data_inspector.py]: Fixed — made all strategy_note assignments conditional on `has_delta` across butterflies, spreads, straddles, strangles, covered_call, protective_put, and single-leg strategies. Each has a fallback note that says "uses OTM% filtering without delta data" to avoid referencing specific delta values when the feature won't work.

File: tests/test_tools_models.py [hunk 23]
```diff
@@ -187,6 +188,169 @@ def test_list_results(self):
...
+        with pytest.raises(ValidationError):
+            DeltaTarget.model_validate(
+                {"target": 0.30, "min": 0.20, "max": 0.40, "extra": 0.1}
+            )
+
+    def test_dumps_to_dict(self):
+        d = DeltaTarget.model_validate({"target": 0.30, "min": 0.20, "max": 0.40})
+        dumped = d.model_dump()
+        assert isinstance(dumped, dict)
+        assert dumped == {"target": 0.30, "min": 0.20, "max": 0.40}
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply tests/test_tools_models.py:238]: Added `test_accepts_boundary_value_one` — validates that `target=1.0, min=0.95, max=1.0` is accepted, explicitly covering the `le=1` boundary.

File: optopsy/ui/tools/_schemas.py [hunk 24]
```diff
@@ -663,15 +664,21 @@ def get_required_option_type(strategy_name: str) -> str | None:
...
         "Analyze a loaded dataset and suggest good starting parameters "
-        "(max_entry_dte, exit_dte) based on the actual DTE "
+        "(max_entry_dte, exit_dte, and per-leg delta targets) based on the actual DTE "
         "and delta distributions. Call this before running a strategy when "
         "the user has not specified explicit parameters, to avoid guessing "
-        "values the data can't satisfy. Returns percentile tables and a "
+        "values the data can't satisfy. Returns percentile tables, per-leg "
+        "delta target recommendations for the given strategy, and a "
         "JSON block of recommended parameter values ready to pass to "
         "run_strategy or scan_strategies."
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/tools/_schemas.py]: Good catch — updated to: `"run_strategy; the DTE-related values can also be reused with scan_strategies."` since scan_strategies doesn't support delta/exit/commission params.

File: optopsy/ui/agent.py [hunk 25]
```diff
@@ -119,26 +119,73 @@
...
+| Strategy | Leg 1 | Leg 2 | Leg 3 | Leg 4 |
+|---|---|---|---|---|
+| Singles (long_calls, short_puts, etc.) | 0.30 (0.20-0.40) | — | — | — |
+| Straddles | 0.50 (0.40-0.60) | 0.50 (0.40-0.60) | — | — |
+| Strangles | 0.30 (0.20-0.40) | 0.30 (0.20-0.40) | — | — |
+| Spreads (long_call_spread, etc.) | 0.50 (0.40-0.60) | 0.10 (0.05-0.20) | — | — |
+| Butterflies | 0.40 (0.30-0.50) | 0.50 (0.40-0.60) | 0.10 (0.05-0.20) | — |
+| Iron condor / reverse | 0.10 (0.05-0.20) | 0.30 (0.20-0.40) | 0.30 (0.20-0.40) | 0.10 (0.05-0.20) |
+| Iron butterfly / reverse | 0.10 (0.05-0.20) | 0.50 (0.40-0.60) | 0.50 (0.40-0.60) | 0.10 (0.05-0.20) |
+| Covered call | 0.80 (0.60-0.95) | 0.30 (0.20-0.40) | — | — |
```
[the remark this answers is not in the document]
Comment (michaelchu) [reply optopsy/ui/agent.py:147]: Added the missing Protective put row to the delta targeting table in the system prompt, matching the Covered call values (0.80 for leg 1, 0.30 for leg 2).
